### PR TITLE
Finish adding getINP()

### LINKED ISSUE #208: Request for support: responsiveness metric
A note to remember that if the test is conclusive, it could be relevant to add to this library the support of the [responsivement metric](https://web.dev/responsiveness/).

### PR DESCRIPTION
**Pull request by philipwalton**
Fixes #208.

This PR builds on top of what @mmocny added in #213 with the following:

- Updates the `observe` module logic to support processing a list of entries rather than each individually.
- Moves the `interactionCount` logic into its own module and rewrites it as a "polyfill" so it won't run if the browser natively supports `interactionCount` (hopefully coming to Chrome soon).
- Sets the `Metric.entries` value to all reported entries from the INP interaction, rather than just the longest. (As discussed we should extend this to all entries in the frame, but we probably need a dedicated API for this first.)
- Updates the `durationThreshold` to 50 as a compromise between performance and more accurate duration values at lower percentiles.
- Adds bfcache support.
- Adds tests

### CODE REVIEW

File: src/getINP.ts [hunk 1]
```diff
@@ -14,118 +14,127 @@
...
+      report();
+    }
   };
 
-  const po = observe('event', entryHandler as PerformanceEntryHandler);
+  const po = observe('event', handleEntries, {
+    // The use of 50 here as a balance between not wanting the callback to
+    // run for too many already fast events (one frame or less), but also
+    // get enough fidelity below the recommended "good" threshold of 200.
+    durationThreshold: 50,
```
Comment (mmocny) [inline src/getINP.ts:114]: Event Timing is rounded to the nearest 8ms.  That makes the duration a nice round number (either 48ms or 56ms). 
 That means that this 50ms durationThreshold is effectively a 56ms duration threshold, which we may want to be explicit about.

TL;DR I think this is a good threshold.  But here are some nuances:

* At 60hz, 56ms is 3.5 frames, which means it can effectively get presented at the same time as other entries with 64ms of latency.
* Conversely, if you were to change to 48ms durationThreshold, you really should probably use 40ms to include entries that take 2.5 frames @ 60hz.
* If you wanted to group entries by frame (i.e. by presentation time) you would need to take startTime + duration and then group entries which are within 8ms windows.  A single shared presentation time can include entries with a wide range of durations.  An entry with less than `durationThreshold` duration can have large processing time.  (This is one reason why we may want a dedicated api)

File: src/getINP.ts [hunk 2]
```diff
@@ -14,118 +14,127 @@
...
+
   report = bindReporter(onReport, metric, reportAllChanges);
 
   if (po) {
     onHidden(() => {
-      po.takeRecords().map(entryHandler as PerformanceEntryHandler);
+      handleEntries(po.takeRecords());
+
+      // If the interaction count shows that there were interactions but
+      // none were captured by the PerformanceObserver, report a latency of 0.
```
Comment (mmocny) [inline src/getINP.ts:124]: Nice!  This cannot happen with the existing polyfill but could with the real version later.

File: src/getINP.ts [hunk 3]
```diff
@@ -14,118 +14,127 @@
...
+      if (metric.value < 0 &&
+          getInteractionCount() - prevInteractionCount > 0) {
```
Comment (mmocny) [inline src/getINP.ts:126]: This seems potentially brittle.  Could you create a simple helper called getInteractionCountForCurrentNavigation() or something, or is this specifically inlined to reduce code size?

File: src/lib/polyfills/interactionCountPolyfill.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+let po: PerformanceObserver | undefined;
+
+/**
+ * Returns the `interactionCount` value using the native API (if available)
+ * or the polyfill estimate in this module.
+ */
+export const getInteractionCount = () => {
+  return po ? interactionCountEstimate : performance.interactionCount || 0;
```
Comment (mmocny) [inline src/lib/polyfills/interactionCountPolyfill.ts:50]: Why `|| 0` at the end?  Defensive?

**Review by mmocny — COMMENTED** [review]
Changes look solid, nice!

Most comments are nits, but I think a few may be actual bugs.

I didn't review or try the unit tests.

File: src/getINP.ts [hunk 5]
```diff
@@ -14,118 +14,127 @@
...
+/**
+ * Takes a performance entry and adds it to the list of worst interactions
+ * if its duration is long enough to make it among the worst. If the
+ * entry is part of an existing interaction, it is merged and the latency
+ * and entries list is updated as needed.
+ */
+const processEntry = (entry: PerformanceEventTiming) => {
+  // Only process the entry if it's possibly one of the ten longest.
+  if (longestInteractions.length < MAX_INTERACTIONS_TO_CONSIDER ||
+      entry.duration > longestInteractions[longestInteractions.length - 1].latency) {
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts:52]: Ahh, good catch. I think we want to keep the optimization (since it contains a `findIndex()` and `sort()`) but add an additional check to see if the entry is part of an existing interaction. I think that covers all cases, right?

File: src/getINP.ts [hunk 6]
```diff
@@ -14,118 +14,127 @@
...
-  const existing = largestINPEntries.findIndex((other) => entry.interactionId == other.interactionId);
-  if (existing >= 0) {
-    // Only replace if this one is actually longer
-    if (entry.duration > largestINPEntries[existing].duration) {
-      largestINPEntries[existing] = entry;
+    // If the interaction already exists, update it. Otherwise create one.
+    if (existingInteractionIndex >= 0) {
+      const interaction = longestInteractions[existingInteractionIndex];
+      interaction.latency = Math.max(interaction.latency, entry.duration);
+      interaction.entries.push(entry);
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts:61]: That was an open question I had (what to sort they by). I think (?) we want to keep them in the order they were dispatched and leave any sorting up to the debug code, but let me know if you disagree and think it's important to have the largest duration first.

File: src/getINP.ts [hunk 7]
```diff
@@ -14,118 +14,127 @@
...
+  latency: number;
+  entries: PerformanceEventTiming[];
 }
 
-function estimateInteractionCount(): number {
-  return (maxKnownInteractionId > 0) ? ((maxKnownInteractionId - minKnownInteractionId) / 7) + 1 : 0;
-}
+// Used to store the interaction count after a bfcache restore, since p98
+// interaction latencies should only consider the current navigation.
+let prevInteractionCount = 0;
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts:33]: > However, I wonder if there are alternative designs that are navigation-id aware?

Yep, had the same thought. I was trying to explain this point when we spoke on Thursday, but clearly I didn't do a good enough job :)

File: src/getINP.ts [hunk 8]
```diff
@@ -14,118 +14,127 @@
...
+      metric.value = inp.latency;
+      metric.entries = inp.entries;
+      report();
+    }
   };
 
-  const po = observe('event', entryHandler as PerformanceEntryHandler);
+  const po = observe('event', handleEntries, {
+    // The use of 50 here as a balance between not wanting the callback to
+    // run for too many already fast events (one frame or less), but also
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts:112]: Yeah, I agree, will update.

File: src/getINP.ts (hunk 2 above)
Comment (philipwalton) [reply src/getINP.ts:124]: It can happen with the existing polyfill because I use `durationThreshold:0` for that.

Comment (philipwalton) [reply src/getINP.ts:124]: It can happen with the existing polyfill because I use `durationThreshold:0` for that.

File: src/getINP.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmocny) [reply src/getINP.ts:61]: Alright, fair enough.

While I think making the first entry the largest would be reasonable given the current implementation -- my ultimate preference is to report all entries that present the same frame such that we can create a summary of where time was spent overall.  (Or even just report the summary as part of the Metric return value).

So leaving it up to debug code sgtm.

File: src/getINP.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmocny) [reply src/getINP.ts:52]: I think so!

File: src/getINP.ts (hunk 3 above)
Comment (philipwalton) [reply src/getINP.ts:126]: I can add that. I agree it would be less brittle.

File: src/lib/polyfills/interactionCountPolyfill.ts (hunk 4 above)
Comment (philipwalton) [reply src/lib/polyfills/interactionCountPolyfill.ts:50]: Yeah. I think there could be a case where the PO errors and the browser doesn't support `interactionCount`. It's unlikely but just in case.

File: src/getINP.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts:61]: I very much agree with that ultimate goal. Let's keep it simple for now and re-evaluate once we have more options there or when we plan to release this as stable.

File: src/getINP.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts]: Done.

File: src/getINP.ts (hunk 1 above)
Comment (philipwalton) [reply src/getINP.ts]: Done.

File: src/getINP.ts (hunk 3 above)
Comment (philipwalton) [reply src/getINP.ts]: Done.

File: src/getINP.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/getINP.ts]: Done.
